tla/consensus/abs.tla

```diff
@@ -37,22 +34,61 @@ Copy(i) ==
         /\ \E l \in 1..(Len(cLogs[j]) - Len(cLogs[i])) : 
                 cLogs' = [cLogs EXCEPT ![i] = @ \o SubSeq(cLogs[j], Len(@) + 1, Len(@) + l)]
 
-\* A node i with the longest log can extend its log upto length k.
-Extend(i, k) ==
+\* A node i with the longest log can non-deterministically extend
+\* its log by any finite number of log entries.  An implementation
+\* may choose a particular number of log entries by which to extend
+\* the log to prevent the leader from racing ahead of the followers.
+Extend(i) ==
     /\ \A j \in Servers : Len(cLogs[j]) \leq Len(cLogs[i])
-    /\ \E l \in 0..(k - Len(cLogs[i])) : 
-        \E s \in [1..l -> Terms] :
+    /\ \E s \in Seq(Terms) :
             cLogs' = [cLogs EXCEPT ![i] = @ \o s]
 
-ExtendToMax(i) == Extend(i, MaxLogLength)
+ExtendAxiom(i) ==
+    \* i has the longest log.
+    /\ \A j \in Servers : Len(cLogs[j]) \leq Len(cLogs[i])
+    \* cLogs remains a function mapping from servers to logs.
+    /\ cLogs' \in [Servers -> Seq(Terms)]
+    \* i *extends* its log
+    /\ IsPrefix(cLogs[i], cLogs'[i])
+    \* The other logs remain the same.
+    /\ \A j \in Servers \ {i} : cLogs'[j] = cLogs[j]
+
+LEMMA ASSUME NEW i \in Servers PROVE
+    ExtendAxiom(i) <=> Extend(i)
+OMITTED 
+
+\* Copy one of the longest logs (from whoever server
+\* has it) and extend it further upto length k. This
+\* is equivalent to  Copy(i) \cdot Extend(i, k)  ,
+\* that TLC cannot handle.
+CopyMaxAndExtend(i) ==
```

Comment:
Removing `CopyMaxAndExtend` in one step breaks the refinement because a node that has not the highest `commitIndex` may move its `commitIndex` past the highest `commitIndex` in a single step.  See https://github.com/microsoft/CCF/pull/6493#issuecomment-2370584056

Reply:
I meant the opposite

Something along the lines of:
```tla
CopyAndOrExtend(i) ==
    \E j \in Servers : 
        /\ Len(cLogs[j]) >= Len(cLogs[i])
        /\ \E l \in Len(cLogs[i]) .. Len(cLogs[j]) : 
          /\ \E s \in Seq(Terms) :
                \* new entry can only be added if i will have a longest length
                /\ l \neq Max({ Len(cLogs[k]) : k \in DOMAIN cLogs}) => s = <<>>
                /\ cLogs' = [cLogs EXCEPT ![i] = @ \o SubSeq(cLogs[j], Len(@) + 1, l) \o s]

Next == \E i \in Servers : CopyAndOrExtend(i)
```

Reply:
In my opinion, this feels a bit too clever, and I find simpler, more straightforward definitions easier to understand.

Reply:
May I mark this resolved?

Reply:
done

Reply:
Does this run as efficiently? I would much prefer a single, 6 line action, to 5 action definitions, two pairs of which are equivalent. It's a lot easier to read!

Reply:
This would still require an associated Axiom action so it would reduce down to 2 not 1

Reply:
Ok, but still substantially shorter?

Reply:
> [...] I would much prefer a single, 6 line action, to 5 action definitions, two pairs of which are equivalent. It's a lot easier to read!

Divide et impera!  The disjunction of `Copy`, `Extend`, and `CopyMaxAndExtend` is a log easier to read than `CopyAndOrExtend`.   (No need to read *Axiom because they are equivalent)

Reply:
Simplex sigillum veri. It's harder to read for me, and it's objectively longer.